include/ibamr/StaggeredStokesIBLevelRelaxationFACOperator.h

```diff
@@ -0,0 +1,677 @@
+                         int coarsest_level_num,
+                         int finest_level_num);
+
+	/*!
+	 * \brief Perform a given number of relaxations on the error.
+	 *
+	 * \param error error vector
+	 * \param residual residual vector
+	 * \param level_num level number
+	 * \param num_sweeps number of sweeps to perform
+	 * \param performing_pre_sweeps boolean value that is true when pre-smoothing sweeps are
+	 *being
+	 *performed
+	 * \param performing_post_sweeps boolean value that is true when post-smoothing sweeps are
+	 *being
+	 *performed
+	 */
+	void smoothError(SAMRAI::solv::SAMRAIVectorReal<NDIM, double>& error,
+					 const SAMRAI::solv::SAMRAIVectorReal<NDIM, double>& residual,
+					 int level_num,
+					 int num_sweeps,
+					 bool performing_pre_sweeps,
+					 bool performing_post_sweeps);
+
+    /*!
+     * \brief Compute hierarchy-dependent data.
+     *
+     * Note that although the vector arguments given to other methods in this
+     * class may not necessarily be the same as those given to this method,
+     * there will be similarities, including:
+     *
+     * - hierarchy configuration (hierarchy pointer and level range)
+     * - number, type and alignment of vector component data
+     * - ghost cell width of data in the solution (or solution-like) vector
+     *
+     * \param solution solution vector u
+     * \param rhs right hand side vector f
+     */
+    void initializeOperatorState(const SAMRAI::solv::SAMRAIVectorReal<NDIM, double>& solution,
+                                 const SAMRAI::solv::SAMRAIVectorReal<NDIM, double>& rhs);
+
+    /*!
+     * \brief Remove all hierarchy-dependent data.
+     *
+     * Remove all hierarchy-dependent data set by initializeOperatorState().
+     *
+     * \see initializeOperatorState
+     */
+    void deallocateOperatorState();
+
+    /*!
+     * \brief Allocate scratch data.
+     */
+    void allocateScratchData();
+
+    /*!
+     * \brief Deallocate scratch data.
+     */
+    void deallocateScratchData();
+
+    //\}
+
+private:
+
+	/*!
+	 * \brief Default constructor.
+	 *
+	 * \note This constructor is not implemented and should not be used.
+	 */
+	StaggeredStokesIBLevelRelaxationFACOperator();
+
+	/*!
+	 * \brief Copy constructor.
+	 *
+	 * \note This constructor is not implemented and should not be used.
+	 *
+	 * \param from The value to copy to this object.
+	 */
+	StaggeredStokesIBLevelRelaxationFACOperator(const StaggeredStokesIBLevelRelaxationFACOperator& from);
+
+	/*!
+	 * \brief Assignment operator.
+	 *
+	 * \note This operator is not implemented and should not be used.
+	 *
+	 * \param that The value to assign to this object.
+	 *
+	 * \return A reference to this object.
+	 */
+	StaggeredStokesIBLevelRelaxationFACOperator& operator=(const StaggeredStokesIBLevelRelaxationFACOperator& that);
+
+    /*!
+     * \name Methods for executing, caching, and resetting communication
+     * schedules.
+     */
+    //\{
+
+    /*!
```

Comment:
I feel so, because I repeating some of the stuff like setComponentsHaveNullspace( )